src/Microsoft.Sbom.Api/Executors/Spdx3SerializationStrategy.cs

```diff
@@ -0,0 +1,135 @@
+        var elementsSpdxIdList = new HashSet<string>();
+
+        WriteContext(sbomConfig);
+
+        sbomConfig.JsonSerializer.StartJsonArray(Constants.SPDXGraphHeaderName);
+
+        // Files section
+        var generateResult = await fileArrayGenerator.GenerateAsync();
+        WriteElementsToSbom(generateResult, elementsSpdxIdList);
+
+        // Packages section
+        var packagesGenerateResult = await packageArrayGenerator.GenerateAsync();
+        generateResult.Errors.AddRange(packagesGenerateResult.Errors);
+        WriteElementsToSbom(packagesGenerateResult, elementsSpdxIdList);
+
+        // External Document Reference section
+        var externalDocumentReferenceGenerateResult = await externalDocumentReferenceGenerator.GenerateAsync();
+        generateResult.Errors.AddRange(externalDocumentReferenceGenerateResult.Errors);
+        WriteElementsToSbom(externalDocumentReferenceGenerateResult, elementsSpdxIdList);
+
+        // Relationships section
+        var relationshipGenerateResult = await relationshipsArrayGenerator.GenerateAsync();
+        generateResult.Errors.AddRange(relationshipGenerateResult.Errors);
+        WriteElementsToSbom(relationshipGenerateResult, elementsSpdxIdList);
+
+        sbomConfig.JsonSerializer.EndJsonArray();
+
+        return generateResult.Errors;
+    }
+
+    private void WriteElementsToSbom(GenerationResult generateResult, HashSet<string> elementsSpdxIdList)
+    {
+        // Write the JSON objects to the SBOM
+        foreach (var serializer in generateResult.SerializerToJsonDocuments.Keys)
+        {
+            var jsonDocuments = generateResult.SerializerToJsonDocuments[serializer];
+            foreach (var jsonDocument in jsonDocuments)
+            {
+                if (jsonDocument.RootElement.ValueKind == JsonValueKind.Object)
+                {
+                    WriteElement(serializer, jsonDocument.RootElement, elementsSpdxIdList);
+                }
+                else
+                {
+                    foreach (var element in jsonDocument.RootElement.EnumerateArray())
+                    {
+                        WriteElement(serializer, element, elementsSpdxIdList);
+                    }
+                }
+            }
+        }
+    }
+
+    private void WriteContext(ISbomConfig sbomConfig)
+    {
+        sbomConfig.JsonSerializer.StartJsonArray(Constants.SPDXContextHeaderName);
+        var document = JsonDocument.Parse(Constants.SPDX3ContextValue);
+        sbomConfig.JsonSerializer.Write(document);
+        sbomConfig.JsonSerializer.EndJsonArray();
+    }
+
+    private void WriteElement(IManifestToolJsonSerializer serializer, JsonElement element, HashSet<string> elementsSpdxIdList)
+    {
+        if (element.TryGetProperty("spdxId", out var spdxIdField))
+        {
+            var spdxId = spdxIdField.GetString();
+
+            if (elementsSpdxIdList.TryGetValue(spdxId, out _))
+            {
+                return;
```

Comment:
Silently continuing is OK if this is a normal case. If it's an unusual case, we might want to log this a debug or verbose level

Reply:
It's not an unusual case to have duplicate elements between different sections. For example, there might be a NoAssertion element generated from `fileArrayGenerator.GenerateAsync()` and `packageArrayGenerator.GenerateAsync()`, but we don't want to add duplicate elements to the SBOM so we silently return if a duplicate is encountered.